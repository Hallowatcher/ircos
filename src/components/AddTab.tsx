
import * as React from 'react';

interface IProps {
  clickAddTab: () => void;
}

const divStyle: React.CSSProperties = {
  padding: '10px',
  margin: '0 2px',
  display: 'flex',
  height: 25,
  alignItems: 'center',
  cursor: 'pointer',
  backgroundColor: 'rgb(241, 196, 15)',
  fontWeight: 800,
  color: 'white',
  boxSizing: 'border-box',
  WebkitUserSelect: 'none'
};

export class AddTab extends React.Component<IProps, any> {
  public render() {
    return (
      <div style={divStyle} onClick={this.props.clickAddTab}>+</div>
    );
  }
}
