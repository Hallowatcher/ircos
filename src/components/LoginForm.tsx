import * as React from 'react';

type Props = {
  submitLogin: (user: string, pass: string) => void
}

type State = {
  user: string,
  pass: string
}

export class LoginForm extends React.Component<Props, State> {

  constructor(props: any) {
    super(props);

    this.state = {
      user: '',
      pass: ''
    }
  }

  inputChanged(event: any) {
    this.setState({
      [event.target.name]: event.target.value
    })
  }

  handleSubmit(event: any) {
    event.preventDefault();

    if (this.state.user !== '' && this.state.pass !== '')
      this.props.submitLogin(this.state.user, this.state.pass);
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit.bind(this)}>
        <input type="text" name="user" value={this.state.user} onChange={this.inputChanged.bind(this)} />
        <input type="password" name="pass" value={this.state.pass} onChange={this.inputChanged.bind(this)} />
        <input type="submit" />
      </form>
    )
  }
}