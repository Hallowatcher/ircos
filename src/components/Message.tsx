import * as React from 'react';
import * as moment from 'moment';

export enum UserType {
  regular,
  moderator,
  self,
  system
}

interface IProps {
  user: string;
  message: string;
  sentDate: moment.Moment;
  userType?: UserType;
}

interface IStyles {
  [key: string]: {
    [key: string]: React.CSSProperties
  } | React.CSSProperties;
}

const styles: IStyles = {
  users: {
    regular: {
      color: '#ffc561',
      fontWeight: 800
    },
    moderator: {
      color: '#BF3434',
      fontWeight: 800
    },
    self: {
      color: '#CECECE',
      fontWeight: 800
    }
  },
  message: {},
  systemMessage: {
    color: '#ADADAD'
  },
  time: {
    color: '#b2b2b2'
  }
};

export class Message extends React.Component<IProps, any> {
  public render() {

    let userStyle = styles.users.regular;
    switch (this.props.userType) {
      case UserType.moderator:
        userStyle = styles.users.moderator;
        break;
      case UserType.self:
        userStyle = styles.users.self;
        break;
    }

    const momentDate = moment(this.props.sentDate);

    // Build date
    const hours = ('0' + momentDate.hours()).slice(-2);
    const minutes = ('0' + momentDate.minutes()).slice(-2);

    let message = (
      <div>
        <span style={styles.time}>[{hours}:{minutes}] </span>
        <span style={userStyle}>{this.props.user}: </span>
        <span style={styles.message}>{this.props.message}</span>
      </div>
    );

    if (this.props.userType === UserType.system) {
      message = (
        <div>
          <span style={styles.systemMessage}>{this.props.message}</span>
        </div>
      );
    }

    return message;
  }
}
