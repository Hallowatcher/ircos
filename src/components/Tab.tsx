
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import {
  DragSource,
  DropTarget,
  DropTargetSpec,
  DragSourceSpec,
  DragSourceCollector,
  DropTargetCollector
} from 'react-dnd';
import { flow } from 'lodash';

interface IProps {
  index: number;
  tabName: string;
  tabClick: (name: string) => void;
  tabMove: (from: number, to: number) => void;
  closeTab: (name: string) => void;
  isActive: boolean;

  // Drag
  connectDragSource?: any;
  connectDropTarget?: any;
  isDragging?: boolean;
  isOver?: boolean;
}

const xStyle: React.CSSProperties = {
  padding: '0 10px',
  margin: '0 0 0 10px',
  backgroundColor: 'rgba(255, 255, 255, 0.15)',
  height: '100%',
  display: 'flex',
  alignItems: 'center',
  fontWeight: 400
};

const divStyle: React.CSSProperties = {
  padding: '0 0 0 10px',
  margin: '0 2px',
  display: 'flex',
  height: 25,
  alignItems: 'center',
  cursor: 'pointer',
  backgroundColor: 'rgba(255, 255, 255, 0.15)',
  fontWeight: 800,
  color: 'white',
  boxSizing: 'border-box',
  WebkitUserSelect: 'none'
};

const divStyleActive: React.CSSProperties = {
  ...divStyle,
  backgroundColor: 'rgba(255, 255, 255, 0.3)'
};

const divStyleDragging: React.CSSProperties = {
  outline: 'white dashed 1px'
};

const tabSource: DragSourceSpec<IProps> = {
  beginDrag(props) {
    return {
      name: props.tabName,
      index: props.index
    };
  }
};

const tabTarget: DropTargetSpec<IProps> = {
  hover(props, monitor, component) {
    const dragIndex = (monitor.getItem() as any).index;
    const hoverIndex = props.index;

    // Don't replace items with themselves
    if (dragIndex === hoverIndex) {
      return;
    }

    // Determine rectangle on screen
    const hoverBoundingRect = ReactDOM.findDOMNode(component).getBoundingClientRect();

    // Get horizontal middle
    const hoverMiddleX = (hoverBoundingRect.right - hoverBoundingRect.left) / 2;

    // Determine mouse position
    const clientOffset = monitor.getClientOffset();

    // Get pixels to the left
    const hoverClientX = clientOffset.x - hoverBoundingRect.left;

    // Only perform the move when the mouse has crossed half of the items height
    // When dragging downwards, only move when the cursor is below 50%
    // When dragging upwards, only move when the cursor is above 50%

    // Dragging left
    if (dragIndex < hoverIndex && hoverClientX < hoverMiddleX) {
      return;
    }

    // Dragging right
    if (dragIndex > hoverIndex && hoverClientX > hoverMiddleX) {
      return;
    }

    // Time to actually perform the action
    props.tabMove(dragIndex, hoverIndex);

    // Note: we're mutating the monitor item here!
    // Generally it's better to avoid mutations,
    // but it's good here for the sake of performance
    // to avoid expensive index searches.
    (monitor.getItem() as any).index = hoverIndex;
  }
};

let collectDrag: DragSourceCollector;
collectDrag = function(connect, monitor) {
  return {
    connectDragSource: connect.dragSource(),
    isDragging: monitor.isDragging()
  };
};

let collectDrop: DropTargetCollector;
collectDrop = function(connect, monitor) {
  return {
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver()
  };
};

class Tab extends React.Component<IProps, null> {
  constructor() {
    super();

    this.handleClick = this.handleClick.bind(this);
    this.handleMiddleClick = this.handleMiddleClick.bind(this);
  }

  public render() {
    const { connectDragSource, connectDropTarget, isDragging, isActive } = this.props;
    return connectDragSource(connectDropTarget(
      <div
        onMouseDown={this.handleMiddleClick}
        style={{...(isDragging ? divStyleDragging : {}), ...(isActive ? divStyleActive : divStyle)}}
      >
        <div
          onClick={this.handleClick}
        >{this.props.tabName}
        </div>
        <div onClick={this.props.closeTab.bind(this, this.props.tabName)} style={xStyle}>&times;</div>
      </div>
    ));
  }
  private handleMiddleClick(event) {
    const type = event.nativeEvent.which;
    if (type !== 2) { return; }
    this.props.closeTab(this.props.tabName);
  }

  private handleClick(event) {
    this.props.tabClick(this.props.tabName);
  }
}

export default flow(
  DragSource('tab', tabSource, collectDrag),
  DropTarget('tab', tabTarget, collectDrop)
)(Tab);
