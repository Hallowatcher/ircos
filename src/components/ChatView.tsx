
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Message, UserType } from './Message';
import * as moment from 'moment';

interface IProps {
  messages: any[];
  selfNick: string;
  userClicked: (user: string) => void;
  style?: React.CSSProperties;
}

export class ChatView extends React.Component<IProps, any> {

  private scrollAtBottom: boolean = true;
  private scrollAtTop: boolean = false;
  private topMessage: any = null;
  private historyChanged: boolean = false;
  private multiplier: number = 1;

  public componentWillUpdate(nextProps: IProps) {
    this.historyChanged = nextProps.messages.length !== this.props.messages.length;
    if (!this.historyChanged || !this.refs.container) { return; }

    const container = this.refs.container as HTMLDivElement;
    const scrollPos = container.scrollTop;
    const scrollBottom = container.scrollHeight - container.clientHeight;
    this.scrollAtBottom = (scrollBottom <= 0) || (scrollPos === scrollBottom);
  }

  public componentDidUpdate() {
    if (this.scrollAtBottom) {
      this.scrollToBottom();
    }
  }

  public render() {
    return (
      <div style={this.props.style} ref="container">
        {this.mapMessages()}
      </div>
    );
  }

  /* istanbul ignore next */
  private scrollToBottom() {
    const container = this.refs.container as HTMLDivElement;
    const scrollHeight = container.scrollHeight;
    const height = container.clientHeight;
    const maxScrollTop = scrollHeight - height;
    ReactDOM.findDOMNode(container).scrollTop = maxScrollTop > 0 ? maxScrollTop : 0;
  }

  private mapMessages() {
    return this.props.messages.map((msg: any, index: number) => {

        // User type check
        let userType = UserType.regular;
        if (msg.name === this.props.selfNick) {
          userType = UserType.self;
        }
        if (msg.name === 'System') {
          userType = UserType.system;
        }

        return (
          <Message
            key={index}
            message={msg.text}
            user={msg.name}
            userType={userType}
            sentDate={moment(msg.date)}
            userClicked={this.props.userClicked}
          />
        );
      });
  }
}
