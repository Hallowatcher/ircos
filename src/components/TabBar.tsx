
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import Tab from './Tab';
import { AddTab } from './AddTab';
import * as FlipMove from 'react-flip-move';

interface IProps {
  tabs: string[];
  currentChannel: string;

  tabClick: (name: string) => void;
  tabMove: (from: number, to: number) => void;
  tabClose: (name: string) => void;
  tabAdd: () => void;
}

const style: React.CSSProperties = {
  flexGrow: 1,
  padding: '0 5px',
  display: 'flex',
  overflowY: 'hidden',
  overflowX: 'auto',
  alignItems: 'flex-end'
};

export class TabBar extends React.Component<IProps, any> {

  public render() {
    const channelMap = this.props.tabs.map((channel, index) =>
    (
      <Tab
        key={channel}
        ref={channel}
        index={index}
        tabName={channel}
        tabClick={this.props.tabClick}
        tabMove={this.props.tabMove}
        closeTab={this.props.tabClose}
        isActive={this.props.currentChannel === channel}
      />
    )
    );

    return (
      <FlipMove style={style} duration={125} easing="ease-out">
        {channelMap}
        <AddTab clickAddTab={this.props.tabAdd} key="add" ref="add" />
      </FlipMove>
    );
  }
}
