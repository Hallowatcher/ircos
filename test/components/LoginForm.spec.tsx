
import * as React from 'react'
import { LoginForm } from '../../src/components/LoginForm'
import * as enzyme from 'enzyme';

describe('AddTab', function () {
  
  let component: enzyme.ShallowWrapper<any, any>
  let submitLogin: any
  beforeEach(function () {
    submitLogin = jest.fn()
    component = enzyme.shallow(<LoginForm submitLogin={submitLogin} error={null} />)
  })

  it('should render', function () {
    expect(component).toMatchSnapshot()
  })

  it('should render with error', function () {
    component = enzyme.shallow(<LoginForm submitLogin={submitLogin} error={'Error'} />)
    expect(component).toMatchSnapshot()
  })

  it('should have an onSubmit prop', function () {
    expect(component.props().onSubmit).toBeDefined()
  })

  it('should submit the right credentials', function () {
    component.find('input[name="user"]').simulate('change', { target: { name: 'user', value: 'foo123' } })
    component.find('input[name="pass"]').simulate('change', { target: { name: 'pass', value: 'bar456' } })
    component.simulate('submit', { preventDefault() {} })
    expect(submitLogin).toBeCalledWith('foo123', 'bar456')
  })

  it('should not submit because credentials aren\'t set', function () {
    component.simulate('submit', { preventDefault() {} })
    expect(submitLogin).toHaveBeenCalledTimes(0)
  })

  it('should not submit credentials because username isn\'t set', function () {
    component.find('input[name="pass"]').simulate('change', { target: { name: 'pass', value: 'bar456' } })
    component.simulate('submit', { preventDefault() {} })
    expect(submitLogin).toHaveBeenCalledTimes(0)
  })

  it('should not submit credentials because password isn\'t set', function () {
    component.find('input[name="user"]').simulate('change', { target: { name: 'user', value: 'foo123' } })
    component.simulate('submit', { preventDefault() {} })
    expect(submitLogin).toHaveBeenCalledTimes(0)
  })

})