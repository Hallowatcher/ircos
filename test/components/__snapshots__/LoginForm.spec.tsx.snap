// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTab should render 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <form
    onSubmit={[Function]}
    style={
        Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "padding": "10px",
          }
    }
>
    <h1>
        ircos
    </h1>
    <input
        name="user"
        onChange={[Function]}
        placeholder="Username"
        style={
            Object {
                "border": 0,
                "borderLeft": "3px solid rgba(0, 0, 0, 0.5)",
                "outline": "none",
                "padding": "4px",
              }
        }
        type="text"
        value=""
    />
    <input
        name="pass"
        onChange={[Function]}
        placeholder="Password"
        style={
            Object {
                "border": 0,
                "borderLeft": "3px solid rgba(0, 0, 0, 0.5)",
                "outline": "none",
                "padding": "4px",
              }
        }
        type="password"
        value=""
    />
    <input
        style={
            Object {
                "border": 0,
                "padding": "4px",
              }
        }
        type="submit"
        value="Login!"
    />
</form>,
  "nodes": Array [
    <form
      onSubmit={[Function]}
      style={
            Object {
                  "alignItems": "center",
                  "display": "flex",
                  "flexDirection": "column",
                  "padding": "10px",
                }
      }
>
      <h1>
            ircos
      </h1>
      <input
            name="user"
            onChange={[Function]}
            placeholder="Username"
            style={
                  Object {
                        "border": 0,
                        "borderLeft": "3px solid rgba(0, 0, 0, 0.5)",
                        "outline": "none",
                        "padding": "4px",
                      }
            }
            type="text"
            value=""
      />
      <input
            name="pass"
            onChange={[Function]}
            placeholder="Password"
            style={
                  Object {
                        "border": 0,
                        "borderLeft": "3px solid rgba(0, 0, 0, 0.5)",
                        "outline": "none",
                        "padding": "4px",
                      }
            }
            type="password"
            value=""
      />
      <input
            style={
                  Object {
                        "border": 0,
                        "padding": "4px",
                      }
            }
            type="submit"
            value="Login!"
      />
</form>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <LoginForm
        error={null}
        submitLogin={[Function]}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": LoginForm {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "handleSubmit": [Function],
        "inputChanged": [Function],
        "props": Object {
          "error": null,
          "submitLogin": [Function],
        },
        "refs": Object {},
        "state": Object {
          "pass": "",
          "user": "",
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <form
          onSubmit={[Function]}
          style={
                    Object {
                              "alignItems": "center",
                              "display": "flex",
                              "flexDirection": "column",
                              "padding": "10px",
                            }
          }
>
          <h1>
                    ircos
          </h1>
          <input
                    name="user"
                    onChange={[Function]}
                    placeholder="Username"
                    style={
                              Object {
                                        "border": 0,
                                        "borderLeft": "3px solid rgba(0, 0, 0, 0.5)",
                                        "outline": "none",
                                        "padding": "4px",
                                      }
                    }
                    type="text"
                    value=""
          />
          <input
                    name="pass"
                    onChange={[Function]}
                    placeholder="Password"
                    style={
                              Object {
                                        "border": 0,
                                        "borderLeft": "3px solid rgba(0, 0, 0, 0.5)",
                                        "outline": "none",
                                        "padding": "4px",
                                      }
                    }
                    type="password"
                    value=""
          />
          <input
                    style={
                              Object {
                                        "border": 0,
                                        "padding": "4px",
                                      }
                    }
                    type="submit"
                    value="Login!"
          />
</form>,
        "_debugID": 2,
        "_renderedOutput": <form
          onSubmit={[Function]}
          style={
                    Object {
                              "alignItems": "center",
                              "display": "flex",
                              "flexDirection": "column",
                              "padding": "10px",
                            }
          }
>
          <h1>
                    ircos
          </h1>
          <input
                    name="user"
                    onChange={[Function]}
                    placeholder="Username"
                    style={
                              Object {
                                        "border": 0,
                                        "borderLeft": "3px solid rgba(0, 0, 0, 0.5)",
                                        "outline": "none",
                                        "padding": "4px",
                                      }
                    }
                    type="text"
                    value=""
          />
          <input
                    name="pass"
                    onChange={[Function]}
                    placeholder="Password"
                    style={
                              Object {
                                        "border": 0,
                                        "borderLeft": "3px solid rgba(0, 0, 0, 0.5)",
                                        "outline": "none",
                                        "padding": "4px",
                                      }
                    }
                    type="password"
                    value=""
          />
          <input
                    style={
                              Object {
                                        "border": 0,
                                        "padding": "4px",
                                      }
                    }
                    type="submit"
                    value="Login!"
          />
</form>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <LoginForm
    error={null}
    submitLogin={[Function]}
/>,
}
`;

exports[`AddTab should render with error 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <form
    onSubmit={[Function]}
    style={
        Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "padding": "10px",
          }
    }
>
    <h1>
        ircos
    </h1>
    <div
        style={
            Object {
                "color": "rgba(255, 0, 0, 0.65)",
                "fontSize": 14,
              }
        }
    >
        Error
    </div>
    <input
        name="user"
        onChange={[Function]}
        placeholder="Username"
        style={
            Object {
                "border": 0,
                "borderLeft": "3px solid rgba(0, 0, 0, 0.5)",
                "outline": "none",
                "padding": "4px",
              }
        }
        type="text"
        value=""
    />
    <input
        name="pass"
        onChange={[Function]}
        placeholder="Password"
        style={
            Object {
                "border": 0,
                "borderLeft": "3px solid rgba(0, 0, 0, 0.5)",
                "outline": "none",
                "padding": "4px",
              }
        }
        type="password"
        value=""
    />
    <input
        style={
            Object {
                "border": 0,
                "padding": "4px",
              }
        }
        type="submit"
        value="Login!"
    />
</form>,
  "nodes": Array [
    <form
      onSubmit={[Function]}
      style={
            Object {
                  "alignItems": "center",
                  "display": "flex",
                  "flexDirection": "column",
                  "padding": "10px",
                }
      }
>
      <h1>
            ircos
      </h1>
      <div
            style={
                  Object {
                        "color": "rgba(255, 0, 0, 0.65)",
                        "fontSize": 14,
                      }
            }
      >
            Error
      </div>
      <input
            name="user"
            onChange={[Function]}
            placeholder="Username"
            style={
                  Object {
                        "border": 0,
                        "borderLeft": "3px solid rgba(0, 0, 0, 0.5)",
                        "outline": "none",
                        "padding": "4px",
                      }
            }
            type="text"
            value=""
      />
      <input
            name="pass"
            onChange={[Function]}
            placeholder="Password"
            style={
                  Object {
                        "border": 0,
                        "borderLeft": "3px solid rgba(0, 0, 0, 0.5)",
                        "outline": "none",
                        "padding": "4px",
                      }
            }
            type="password"
            value=""
      />
      <input
            style={
                  Object {
                        "border": 0,
                        "padding": "4px",
                      }
            }
            type="submit"
            value="Login!"
      />
</form>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <LoginForm
        error="Error"
        submitLogin={[Function]}
/>,
      "_debugID": 5,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": LoginForm {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "handleSubmit": [Function],
        "inputChanged": [Function],
        "props": Object {
          "error": "Error",
          "submitLogin": [Function],
        },
        "refs": Object {},
        "state": Object {
          "pass": "",
          "user": "",
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <form
          onSubmit={[Function]}
          style={
                    Object {
                              "alignItems": "center",
                              "display": "flex",
                              "flexDirection": "column",
                              "padding": "10px",
                            }
          }
>
          <h1>
                    ircos
          </h1>
          <div
                    style={
                              Object {
                                        "color": "rgba(255, 0, 0, 0.65)",
                                        "fontSize": 14,
                                      }
                    }
          >
                    Error
          </div>
          <input
                    name="user"
                    onChange={[Function]}
                    placeholder="Username"
                    style={
                              Object {
                                        "border": 0,
                                        "borderLeft": "3px solid rgba(0, 0, 0, 0.5)",
                                        "outline": "none",
                                        "padding": "4px",
                                      }
                    }
                    type="text"
                    value=""
          />
          <input
                    name="pass"
                    onChange={[Function]}
                    placeholder="Password"
                    style={
                              Object {
                                        "border": 0,
                                        "borderLeft": "3px solid rgba(0, 0, 0, 0.5)",
                                        "outline": "none",
                                        "padding": "4px",
                                      }
                    }
                    type="password"
                    value=""
          />
          <input
                    style={
                              Object {
                                        "border": 0,
                                        "padding": "4px",
                                      }
                    }
                    type="submit"
                    value="Login!"
          />
</form>,
        "_debugID": 6,
        "_renderedOutput": <form
          onSubmit={[Function]}
          style={
                    Object {
                              "alignItems": "center",
                              "display": "flex",
                              "flexDirection": "column",
                              "padding": "10px",
                            }
          }
>
          <h1>
                    ircos
          </h1>
          <div
                    style={
                              Object {
                                        "color": "rgba(255, 0, 0, 0.65)",
                                        "fontSize": 14,
                                      }
                    }
          >
                    Error
          </div>
          <input
                    name="user"
                    onChange={[Function]}
                    placeholder="Username"
                    style={
                              Object {
                                        "border": 0,
                                        "borderLeft": "3px solid rgba(0, 0, 0, 0.5)",
                                        "outline": "none",
                                        "padding": "4px",
                                      }
                    }
                    type="text"
                    value=""
          />
          <input
                    name="pass"
                    onChange={[Function]}
                    placeholder="Password"
                    style={
                              Object {
                                        "border": 0,
                                        "borderLeft": "3px solid rgba(0, 0, 0, 0.5)",
                                        "outline": "none",
                                        "padding": "4px",
                                      }
                    }
                    type="password"
                    value=""
          />
          <input
                    style={
                              Object {
                                        "border": 0,
                                        "padding": "4px",
                                      }
                    }
                    type="submit"
                    value="Login!"
          />
</form>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <LoginForm
    error="Error"
    submitLogin={[Function]}
/>,
}
`;
